### PR TITLE
Add support for N-dimensional arrays in imgradients

### PR DESCRIPTION
**Pull request by juliohm**
Hi Tim,

I accidentally deleted the previous pull request during my Git adventures, opening a new one with corrections.

### CODE REVIEW

**Comment by juliohm** [comment]
A new error is showing in Travis, could you please comment on how to fix it?

File: src/edge.jl [hunk 1]
```diff
@@ -109,40 +109,121 @@ function ando5_sep()
...
+
+    result
 end
 
+@deprecate imgradients(img::AbstractArray, method::AbstractString, border::AbstractString) imgradients(img; method=method, border=border)
+
 function imgradients{T<:Color}(img::AbstractArray{T}, method::AbstractString="ando3", border::AbstractString="replicate")
     # Remove Color information
-    imgradients(reinterpret(eltype(eltype(img)), img), method, border)
+    imgradients(reinterpret(eltype(eltype(img)), img), method=method, border=border)
```
Comment (timholy) [inline src/edge.jl]: Your Travis error is probably due to this method not being called anymore (note this one isn't using kwargs).

Comment (juliohm) [inline src/edge.jl]: Hi Tim, how can I fix this issue?

Comment (timholy) [inline src/edge.jl]: Change the signature to be the same as your other methods (i.e., insert a semicolon after the `img` argument).

**Comment by juliohm** [comment]
Hi Tim, we are almost there. The last error is related to RGB channels. What is the expected behavior for imgradients on the channels? Should it be applied in 2D for each channel separately and the result put together in this 3xmxn convention?

Also, we have another issue with the non-separable kernels, I couldn't figure out a way to tweak the code in order to incorporate ando4, and ando5 for example. Should we have an extra imgradients2D supporting these kernels instead?

**Comment by timholy** [comment]
> The last error is related to RGB channels. What is the expected behavior for imgradients on the channels? Should it be applied in 2D for each channel separately and the result put together in this 3xmxn convention?

The tests are basically what define the behavior we expect/enforce (that doesn't mean it can't change in the future, but this would be a breaking change, not a minor update). That seems to be [what's expected](https://github.com/timholy/Images.jl/blob/f629d05c981f1d676d6ab1d4e05543f181c0fafb/test/edge.jl#L227).

> Also, we have another issue with the non-separable kernels, I couldn't figure out a way to tweak the code in order to incorporate ando4, and ando5 for example. Should we have an extra imgradients2D supporting these kernels instead?

Or allow them when calling `imgradients` on 2d images, but not otherwise.

**Comment by juliohm** [comment]
I tried to rewrite the function to something like:

```julia
function imgradients{T<:Color}(img::AbstractArray{T}; method::AbstractString="ando3", border::AbstractString="replicate")
    # Remove Color information
    rawimg = reinterpret(eltype(eltype(img)), img)
    Rx, Ry = imgradients(rawimg[1,:,:], method=method, border=border)
    Gx, Gy = imgradients(rawimg[2,:,:], method=method, border=border)
    Bx, By = imgradients(rawimg[3,:,:], method=method, border=border)

    gradx = cat(1, Rx, Gx, Bx)
    grady = cat(1, Ry, Gy, By)

    gradx, grady
end
```
without success. I just need to know how the filtering looks like when one of the dimensions means a channel.

Regarding the non-separable kernels, I will add a big if-else for 2D images...

**Comment by timholy** [comment]
Couldn't you have just left the old one there and add the semicolon? I think you don't need to do all that, just reshape the kernel so it's of size 1 along the first dimension.

**Comment by timholy** [comment]
When in doubt, don't forget: you can always commit or stash your current work, check out master, and run a quick test to see how the current code works. Then go back to yours and make adjustments.

**Comment by juliohm** [comment]
Yes Tim, that is probably the best way to handle these special cases. I will just revive the old implementation, call it _imgradients2D() and put a big if-else in the beginning of imgradients.

**Comment by juliohm** [comment]
Hi Tim, I am falling back to the old implementation for 2D images, but I am still getting similar errors, could you please take a quick look when you have time? I will squash all these commits after the tests pass, sorry for the noise, getting more and more used to Git control.

**Comment by juliohm** [comment]
This is moved to `ImagesFiltering.jl`.

**Pull request closed without merging**